File: app/src/main/java/kr/djspi/pipe01/ViewActivity.java
package kr.djspi.pipe01;

import android.content.Intent;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.design.widget.TabLayout.OnTabSelectedListener;
import android.support.design.widget.TabLayout.Tab;
import android.support.design.widget.TabLayout.TabLayoutOnPageChangeListener;
import android.support.v4.view.ViewPager;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.helloelliote.json.Json;

import org.jetbrains.annotations.NotNull;

import java.io.Serializable;
import java.util.ArrayList;

import kr.djspi.pipe01.dto.Entry;
import kr.djspi.pipe01.dto.SpiLocation;
import kr.djspi.pipe01.tab.OnRecordListener;
import kr.djspi.pipe01.tab.TabAdapter;

import static android.content.Intent.ACTION_DIAL;
import static android.net.Uri.parse;
import static android.text.Html.fromHtml;
import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static java.util.Objects.requireNonNull;
import static kr.djspi.pipe01.Const.REQUEST_CODE_MAP;
import static kr.djspi.pipe01.Const.RESULT_FAIL;
import static kr.djspi.pipe01.Const.RESULT_PASS;

public class ViewActivity extends BaseActivity implements Serializable, OnRecordListener {

    private static final String TAG = ViewActivity.class.getSimpleName();
    private int pipeIndex;
    private JsonObject jsonObject;
    private ArrayList<Entry> previewEntries;
    /**
     * 아래의 변수들은 내부 클래스에서도 참조하는 변수로, private 선언하지 않는다.
     */
    ViewPager viewPager;

    @Override
    @SuppressWarnings("unchecked")
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        previewEntries = new ArrayList<>();

        final Intent intent = getIntent();
        if (intent != null) {
            String jsonString = intent.getStringExtra("PipeView");
            if (jsonString != null) {
                jsonObject = new JsonParser().parse(jsonString).getAsJsonObject();
            }

            Serializable serializable = intent.getSerializableExtra("RegisterPreview");
            pipeIndex = intent.getIntExtra("PipeIndex", 0);
            String fHorizontal = intent.getStringExtra("fHorizontal");
            String fVertical = intent.getStringExtra("fVertical");
            if (serializable instanceof ArrayList<?>) {
                previewEntries = (ArrayList<Entry>) serializable;
                jsonObject = new JsonObject();
                jsonObject = parseEntry(previewEntries, pipeIndex, fHorizontal, fVertical); // 단일형 index 는 항상 0
            }
        }
        setContentView(R.layout.activity_pipe_view);
    }

    @Override
    public void setContentView(int layoutResID) {
        super.setContentView(layoutResID);

        // (탭) 사용자의 좌우 스와이프 이벤트를 처리가능하게끔 어댑터를 사용
        TabLayout tabLayout = findViewById(R.id.tabs);
        if (previewEntries.isEmpty()) {
            tabLayout.removeTab(requireNonNull(tabLayout.getTabAt(3)));
        }
        final TabAdapter tabAdapter = new TabAdapter(getSupportFragmentManager(), tabLayout.getTabCount());
        viewPager = findViewById(R.id.container);
        viewPager.setAdapter(tabAdapter);
        viewPager.addOnPageChangeListener(new TabLayoutOnPageChangeListener(tabLayout));
        tabLayout.addOnTabSelectedListener(new TabSelected());

        LinearLayout linearLayout = (LinearLayout) tabLayout.getChildAt(0);
        linearLayout.setShowDividers(LinearLayout.SHOW_DIVIDER_MIDDLE);
        GradientDrawable drawable = new GradientDrawable();
        drawable.setColor(getResources().getColor(R.color.yellow));
        drawable.setSize(8, 1);
        linearLayout.setDividerDrawable(drawable);

        setToolbarTitle("");

        setSpiIdInfo();
        setSuperviseInfo();
        setConstructionInfo();
    }

    @Override
    void setToolbarTitle(String string) {
        if (string != null) {
            toolbar.setTitle(String.format(getString(R.string.app_title_alt), Json.s(jsonObject, "pipe"), ""));
        }
    }

    private void setSpiIdInfo() {
        TextView id = findViewById(R.id.txt_id);
        try {
            String idInt = Json.s(jsonObject, "spi_id");
            id.setText(fromHtml(getString(R.string.nfc_info_id, idInt)));
        } catch (RuntimeException e) {
            id.setVisibility(GONE);
        }
    }

    private void setSuperviseInfo() {
        TextView name = findViewById(R.id.txt_company);
        TextView contact = findViewById(R.id.txt_contact);
        try {
            String superviseName = Json.s(jsonObject, "supervise");
            name.setText(fromHtml(getString(R.string.nfc_info_company, superviseName)));

            String superviseContact = Json.s(jsonObject, "supervise_contact");
            contact.setText(fromHtml(getString(R.string.nfc_info_company_contact, superviseContact)));
            contact.setOnClickListener(v -> startActivity(new Intent(ACTION_DIAL, parse("tel:" + superviseContact))));
        } catch (RuntimeException e) {
            name.setVisibility(GONE);
            contact.setVisibility(GONE);
        }
    }

    private void setConstructionInfo() {
        TextView name = findViewById(R.id.txt_construction);
        name.setVisibility(GONE);
        try {
            String constructionName = Json.s(jsonObject, "construction");
            String constructionContact = Json.s(jsonObject, "construction_contact");
            if (constructionName.length() > 0 || constructionContact.length() > 0) {
                name.setVisibility(VISIBLE);
                name.setText(fromHtml(getString(R.string.nfc_info_construction, constructionName, constructionContact)));
                if (constructionContact.length() > 0) {
                    name.setOnClickListener(v -> startActivity(new Intent(ACTION_DIAL, parse("tel:" + constructionContact))));
                }
            }
        } catch (RuntimeException e) {
            name.setVisibility(GONE);
        }
    }

    @Override
    public JsonObject getJsonObjectRecord() {
        return jsonObject;
    }

    @Override
    public void onRecord(String tag, int result) {
        switch (result) {
            case RESULT_PASS:
                startActivityForResult(new Intent(this, SpiLocationActivity.class)
                        .setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP), REQUEST_CODE_MAP);
                break;
            case RESULT_FAIL:
                break;
            default:
                break;
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK) {
            switch (requestCode) {
                case REQUEST_CODE_MAP:
                    double[] locations = data.getDoubleArrayExtra("locations");
                    Entry currentEntry = previewEntries.get(pipeIndex);
                    SpiLocation location = currentEntry.getSpi_location();
                    location.setLatitude(locations[0]);
                    location.setLongitude(locations[1]);
                    location.setCount(0);
                    currentEntry.setSpi_location(location);
                    previewEntries.set(pipeIndex, currentEntry);
                    startActivity(new Intent(this, SpiPostActivity.class)
                            .setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
                            .putExtra("entry", previewEntries));
                    break;
                default:
                    break;
            }
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        nfcUtil.onPause();
    }

    private class TabSelected implements OnTabSelectedListener {

        @Override
        public void onTabSelected(@NonNull Tab tab) {
            viewPager.setCurrentItem(tab.getPosition());
            LinearLayout layout = findViewById(R.id.lay_bottom);
            if (tab.getPosition() == 3) layout.setVisibility(GONE);
            else layout.setVisibility(VISIBLE);
        }

        @Override
        public void onTabUnselected(Tab tab) {

        }

        @Override
        public void onTabReselected(Tab tab) {

        }
    }

    @SuppressWarnings("SameParameterValue")
    private static JsonObject parseEntry(@NotNull ArrayList entries, int index, String... strings) {
        return ((Entry) entries.get(index)).parseToSingleJsonObject(strings);
    }
}

File: app/src/main/java/kr/djspi/pipe01/ViewActivity.kt
package kr.djspi.pipe01

import android.content.Intent
import android.content.Intent.ACTION_DIAL
import android.graphics.drawable.GradientDrawable
import android.net.Uri
import android.net.Uri.parse
import android.os.Bundle
import android.view.View
import android.widget.LinearLayout
import androidx.appcompat.widget.Toolbar
import androidx.viewpager.widget.ViewPager
import com.google.android.material.tabs.TabLayout
import com.google.gson.JsonObject
import com.google.gson.JsonParser.parseString
import kotlinx.android.synthetic.main.activity_base.*
import kotlinx.android.synthetic.main.activity_pipe_view.*
import kr.djspi.pipe01.Const.REQUEST_MAP
import kr.djspi.pipe01.Const.RESULT_PASS
import kr.djspi.pipe01.dto.Entry
import kr.djspi.pipe01.dto.Entry.Companion.parseEntry
import kr.djspi.pipe01.dto.SpiPhotoObject
import kr.djspi.pipe01.tab.InfoTab
import kr.djspi.pipe01.tab.OnRecordListener
import kr.djspi.pipe01.tab.TabAdapter
import kr.djspi.pipe01.util.fromHtml
import kr.djspi.pipe01.util.onNewIntentIgnore
import kr.djspi.pipe01.util.onPauseNfc
import kr.djspi.pipe01.util.onResumeNfc
import java.io.Serializable

class ViewActivity : BaseActivity(), Serializable, OnRecordListener {

    private var pipeIndex: Int = 0
    private var photoObject: SpiPhotoObject? = null
    private var previewEntries: ArrayList<Entry>? = null
    private var viewPager: ViewPager? = null
    private lateinit var jsonObj: JsonObject

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        intent?.let {
            val jsonString = intent.getStringExtra("PipeView")
            if (jsonString != "Register") {
                jsonObj = parseString(jsonString).asJsonObject
            }

            val preview = it.getSerializableExtra("RegisterPreview")
            val fHorizontal = it.getStringExtra("fHorizontal")
            val fVertical = it.getStringExtra("fVertical")
            pipeIndex = it.getIntExtra("PipeIndex", 0)
            if (preview is ArrayList<*>) {
                previewEntries = preview as ArrayList<Entry>
                jsonObj = parseEntry(previewEntries!!, pipeIndex, fHorizontal, fVertical)
            }
        }

        val classSerializable = intent.getSerializableExtra("PhotoObj")
        if (classSerializable != null) {
            photoObject = classSerializable as SpiPhotoObject
        }

        setContentView(R.layout.activity_pipe_view)

        runOnUiThread {
            setSpiIdInfo()
            setSuperviseInfo()
            setConstructionInfo()
        }
    }

    override fun setContentView(layoutResID: Int) {
        super.setContentView(layoutResID)
        toolbar.title = "SPI ${jsonObj["pipe"].asString}"
        setTabLayout()
    }

    override fun setSupportActionBar(toolbar: Toolbar?) {
        super.setSupportActionBar(toolbar)
        super.nmapFind.text = getString(R.string.btn_current_spi)
        super.nmapFind.apply {
            setOnClickListener {
                startActivity(
                    Intent(context, NaverMapActivity::class.java)
                        .setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
                        .putExtra("isSpiLocation", true)
                        .putExtra("spi_latitude", jsonObj["spi_latitude"].asDouble)
                        .putExtra("spi_longitude", jsonObj["spi_longitude"].asDouble)
                )
            }
        }
    }

    private fun setTabLayout() {
        if (previewEntries == null) {
            tabs.removeTab(tabs.getTabAt(3)!!)
        }
        viewPager = findViewById(R.id.container)
        viewPager?.let {
            it.adapter = TabAdapter(supportFragmentManager, tabs.tabCount)
            it.addOnPageChangeListener(TabLayout.TabLayoutOnPageChangeListener(tabs))
        }
        tabs.addOnTabSelectedListener(TabSelected())
        val linearLayout: LinearLayout = tabs.getChildAt(0) as LinearLayout
        linearLayout.apply {
            showDividers = LinearLayout.SHOW_DIVIDER_MIDDLE
            dividerDrawable = GradientDrawable().apply {
                setColor(resources.getColor(R.color.yellow, null))
                setSize(8, 1)
            }
        }
    }

    private fun setSpiIdInfo() {
        if (jsonObj["id"] == null) {
            txt_id.visibility = View.GONE
        } else {
            txt_id.text = fromHtml(getString(R.string.nfc_info_id, jsonObj["id"].asString))
        }
    }

    private fun setSuperviseInfo() {
        if (jsonObj["supervise"] == null) {
            txt_company.visibility = View.GONE
            txt_contact.visibility = View.GONE
        } else {
            txt_company.text =
                fromHtml(getString(R.string.info_company, jsonObj["supervise"].asString))
            txt_contact.text =
                fromHtml(getString(R.string.info_contact, jsonObj["supervise_contact"].asString))
            txt_contact.setOnClickListener {
                startActivity(
                    Intent(
                        ACTION_DIAL,
                        parse("tel:${jsonObj["supervise_contact"].asString}")
                    )
                )
            }
        }
    }

    private fun setConstructionInfo() {
        if (jsonObj["construction"] == null) {
            txt_construction.visibility = View.GONE
        } else {
            val construction = jsonObj["construction"].asString
            val constructionContact = jsonObj["construction_contact"].asString
            if (construction.isNotEmpty() || constructionContact.isNotEmpty()) {
                txt_construction.visibility = View.VISIBLE
                txt_construction.text = fromHtml(
                    getString(
                        R.string.nfc_info_construction,
                        construction,
                        constructionContact
                    )
                )
                if (constructionContact.isNotEmpty()) {
                    txt_construction.setOnClickListener {
                        startActivity(Intent(ACTION_DIAL, parse("tel:$constructionContact")))
                    }
                }
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == RESULT_OK && data != null) {
            if (requestCode == REQUEST_MAP) {
//                val locations: DoubleArray = data.getDoubleArrayExtra("locations")!!
//                spiLocation?.latitude = locations[0]
//                spiLocation?.longitude = locations[1]

                val spiLocations =
                    data.getSerializableExtra("spiLocations") as HashMap<*, *>
                val pipeLocations =
                    data.getSerializableExtra("pipeLocations") as HashMap<*, *>
                val currentEntry = previewEntries!![pipeIndex]

                if (spiLocations.isNotEmpty()) {
                    val spiLocation = currentEntry.spi_location
                    spiLocation?.latitude = spiLocations["latitude"] as Double?
                    spiLocation?.longitude = spiLocations["longitude"] as Double?
                    spiLocation?.coordinate_x = spiLocations["coordinate_x"] as Double?
                    spiLocation?.coordinate_y = spiLocations["coordinate_y"] as Double?
                    spiLocation?.origin = spiLocations["origin"] as String?
                    spiLocation?.count = 0
                    currentEntry.spi_location = spiLocation
                }

                if (pipeLocations.isNotEmpty()) {
                    val pipeLocation = currentEntry.pipe_location
                    pipeLocation?.latitude = pipeLocations["latitude"] as Double?
                    pipeLocation?.longitude = pipeLocations["longitude"] as Double?
                    pipeLocation?.coordinate_x = pipeLocations["coordinate_x"] as Double?
                    pipeLocation?.coordinate_y = pipeLocations["coordinate_y"] as Double?
                    pipeLocation?.origin = pipeLocations["origin"] as String?
                    pipeLocation?.count = 0
                    currentEntry.pipe_location = pipeLocation
                }

                previewEntries!![pipeIndex] = currentEntry
                startActivity(
                    Intent(this, SpiPostActivity::class.java)
                        .setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
                        .putExtra("entry", previewEntries)
                        .putExtra("Photos", photoObject)
                )
            }
        }
    }

    override val jsonObject: JsonObject
        get() = jsonObj
    override val uri: Uri?
        get() = if (photoObject == null) null else photoObject!!.getUri()


    override fun onRecord(tag: String, result: Int) {
        when (result) {
            RESULT_PASS -> {
                this.startActivityForResult(
                    Intent(this, SpiLocationActivity::class.java)
                        .setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
                        .putExtra("pipeType", jsonObj["pipe"].asString)
                        .putExtra("horizontal", jsonObj["horizontal"].asDouble)
                        .putExtra("vertical", jsonObj["vertical"].asDouble)
                        .putExtra("hDirection", InfoTab.getHorizontalDirection(jsonObj))
                        .putExtra("vDirection", InfoTab.getVerticalDirection(jsonObj)), REQUEST_MAP
                )
            }
        }
    }

    override fun onResume() {
        super.onResume()
        onResumeNfc()
    }

    override fun onPause() {
        super.onPause()
        onPauseNfc()
    }

    override fun onDestroy() {
        super.onDestroy()
        viewPager = null
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        onNewIntentIgnore()
    }

    private inner class TabSelected : TabLayout.OnTabSelectedListener {
        override fun onTabSelected(tab: TabLayout.Tab) {
            viewPager?.currentItem = tab.position
            if (tab.position == 3) lay_bottom.visibility = View.GONE
            else lay_bottom.visibility = View.VISIBLE
        }

        override fun onTabReselected(tab: TabLayout.Tab?) {
        }

        override fun onTabUnselected(tab: TabLayout.Tab?) {
        }
    }
}
